unpack.c

```diff
@@ -50,26 +51,36 @@ ExtractedArchive* extract_archive(uint8_t* inputData, size_t inputSize, size_t*
         archive_read_free(archive);
         return result;
     }
+    files = malloc(sizeof(FileData) * files_struct_length);
 
     while (archive_read_next_header(archive, &entry) == ARCHIVE_OK) {
         const char* filename =  decompressionOnly ? "decompression.json": archive_entry_pathname(entry);
         size_t entrySize = decompressionOnly ? inputSize: archive_entry_size(entry);
-        files= realloc(files, sizeof(FileData) * (files_count + 1));
-        if (!files) {
-            archive_read_free(archive);
-            result->status = 0;
-            snprintf(result->error_message, sizeof(result->error_message), "Memory allocation error for file data.");
-            return result;
+        if (files_count + 1 > files_struct_length) {
+            files_struct_length *= 2; // double the length
+            FileData* oldfiles = files;
+            files= realloc(files, sizeof(FileData) * files_struct_length);
+            if (!files) {
+                archive_read_free(archive);
+                result->status = 0;
+                result->fileCount = files_count;
+                result->files = oldfiles; // otherwise memory is lost, alternatively also everything can be freed.
+                snprintf(result->error_message, sizeof(result->error_message), "Memory allocation error for file data.");
```

Comment:
I was a bit puzzled here. In the memory allocation error, the file structure was not put into the result, so the memory was lost. The JS side will call free_extracted_archive anyway, so I wonder why no cleanup code was executed in these branches compared to an unsuccessful archive_read call. 
Avoiding duplicated cleanup code may be a good idea.